### ISSUE TITLE
Removing a layer causes map.loaded() to return false until map is moved.

### ISSUE THREAD

**Issue by dpieri** [mapbox/mapbox-gl-js#2936]
**mapbox-gl-js version**:0.21.0

**jsBin**: http://jsbin.com/fewegikejo/edit?html,console,output

### Steps to Trigger Behavior

 1. Check that map.loaded() returns true
 2. Remove any layer from the map
 3. Check that map.loaded() returns false a few times
 4. Pan or zoom the map
 5. See that map.loaded() now returns true

**Comment by dpieri** [comment]
I've also noticed in our own app that a map initialized at a high zoom will report `map.loaded()` to be `false` until you zoom out and then back in again. I know the next release will have some significant changes to the source interface. Should I assume that these issues will be resolved in that release, or do I need to find a way to hack around this? Parts of our app wait until the map is loaded to reveal themselves, so this is a pretty big issue.

**Comment by anandthakker** [comment]
This is happening because in the `_update` => `_render` triggered by `removeLayer`, [this line](https://github.com/mapbox/mapbox-gl-js/blob/8b293c4bea7706456ddb74d6a47e41a0dd0a76fb/js/ui/map.js#L1007) is firing a style change event which (_synchronously) resets `_styleDirty`.

Seems possible that the fix might be as simple as adding `this._styleDirty` to this line: https://github.com/mapbox/mapbox-gl-js/blob/8b293c4bea7706456ddb74d6a47e41a0dd0a76fb/js/ui/map.js#L1038 -- but this is touching some possibly sensitive render-loop logic -- @lucaswoj can you give this a sanity check?

**Comment by lucaswoj** [comment]
The unit test for this bugfix is not working consistently. More investigation is needed.

**Comment by anandthakker** [comment]
> The unit test for this bugfix is not working consistently

@lucaswoj Can you describe the inconsistency you observed?

**Comment by lucaswoj** [comment]
It has been failing about 50% of the time on `master`. I'm not sure if this is a problem with the unit test or the bug fix. In order to unblock the v0.22.0 release I reverted this PR.

**Comment by anandthakker** [comment]
Gotcha.  I admit I was guessing a little bit at the right way to control timing for that unit test; seems like a pretty good chance I got it wrong.  I'll check it out.

**Comment by dpieri** [comment]
Shouldn't `Fix bug causing Map#loaded to return false after removing a layer #2936` be removed from the CHANGELOG notes for 0.22.0 now that this has been reopened?

**Comment by lucaswoj** [comment]
@dpieri https://github.com/mapbox/mapbox-gl-js/commit/bbcd17c299c6c11f0934d16d312d25aec0aec648#diff-4ac32a78649ca5bdd8e0ba38b7006a1eL56

**Comment by dpieri** [comment]
The jsBin test case I included in this original issue is fixed by 0.22.1, but there are still cases where map.loaded() reports false until the map is moved. For example, if you visit [here where I just upgraded to 0.22.1](https://hivemapper.com/state/california/census-designated-place/west-menlo-park-57a276cd585b4ed17dce741a) the polygon for West Menlo Park only shows up after map.loaded() returns true. You can type map.loaded() in the javascript console, or just zoom in and out until the polygon shows up. If you do nothing map.loaded() reports false forever.

Are there more changes coming in 0.23 or should I open up a new issue for these other cases?
